utils/pce-mkcd/pce-mkcd.cc

```diff
@@ -0,0 +1,668 @@
+          strtab_offset = int_at(elf_data, shdr + ELF32_SHDR_OFFSET);
+          strtab_size = int_at(elf_data, shdr + ELF32_SHDR_SIZE);
+        } break;
+        case SHT_SYMTAB: {
+          symtab_offset = int_at(elf_data, shdr + ELF32_SHDR_OFFSET);
+          symtab_size = int_at(elf_data, shdr + ELF32_SHDR_ENTSIZE);
+          symtab_count = int_at(elf_data, shdr + ELF32_SHDR_SIZE);
+          if (symtab_count == 0 || symtab_size == 0 || (symtab_count % symtab_size) != 0) {
+            error(1, "Unsupported layout of .symtab!");
+          }
+          symtab_count /= symtab_size;
+        } break;
+      }
+    }
+  }
+
+  virtual void write(Disc &disc, std::ofstream &out) override {
+    std::vector<char> data(sectors() * SECTOR_SIZE);
+    uint8_t mpr_mapping[5] = {0, 1, 2, 3, 4};
+    size_t start = out.tellp();
+
+    memset(&data[0], 0, data.size());
+
+    // Read Phdr contents.
+    for (auto i = 0; i < phdr_count; i++) {
+      uint32_t phdr = phdr_offset + (phdr_size * i);
+      if (int_at(elf_data, phdr + ELF32_PHDR_TYPE) == PT_LOAD) {
+        uint32_t offset = int_at(elf_data, phdr + ELF32_PHDR_OFFSET);
+        uint32_t filesz = int_at(elf_data, phdr + ELF32_PHDR_FILESZ);
+        uint32_t addr = int_at(elf_data, phdr + ELF32_PHDR_PADDR);
+        uint32_t file_offset = addr_to_file_offset(addr);
+        memcpy(&data[file_offset], &elf_data[offset], filesz);
+
+        uint8_t mpr_bank = (addr >> 16) & 0xFF;
+        uint8_t mpr_index = (addr >> 13) & 0x7;
+        if (mpr_index >= 2 && mpr_index <= 6 && mpr_bank >= 0x80 && mpr_bank <= 0x87) {
+          mpr_mapping[mpr_index - 2] = mpr_bank - 0x80;
+        }
+      }
+    }
+
+    // Apply Rela contents.
+    for (auto i = 0; i < shdr_count; i++) {
+      uint32_t shdr = shdr_offset + (shdr_size * i);
+      if (int_at(elf_data, shdr + ELF32_SHDR_TYPE) == SHT_RELA) {
+        uint32_t rela_offset = int_at(elf_data, shdr + ELF32_SHDR_OFFSET);
+        uint32_t rela_size = int_at(elf_data, shdr + ELF32_SHDR_ENTSIZE);
+        uint32_t rela_count = int_at(elf_data, shdr + ELF32_SHDR_SIZE);
+        for (auto j = 0; j < rela_count; j += rela_size) {
+          uint32_t rela = rela_offset + j;
+          uint32_t r_addr = int_at(elf_data, rela + ELF32_RELA_OFFSET);
+          uint32_t r_file_offset = addr_to_file_offset(r_addr);
+          uint32_t r_sym_index = ELF32_R_SYM(int_at(elf_data, rela + ELF32_RELA_INFO));
+          uint32_t r_sym_type = ELF32_R_TYPE(int_at(elf_data, rela + ELF32_RELA_INFO));
+          uint32_t r_sym = symtab_offset + r_sym_index * symtab_size;
+          std::string r_sym_name = strtab_name(int_at(elf_data, r_sym + ELF32_SYM_NAME));
+          uint32_t r_sym_value = int_at(elf_data, r_sym + ELF32_SYM_VALUE);
+          uint16_t r_sym_shndx = short_at(elf_data, r_sym + ELF32_SYM_SHNDX);
+
+          // All resolved symbols are relocated - look only for unresolved symbols.
+          if (starts_with(r_sym_name, cd_symbol_prefix)) {
```

Comment:
Just thinking out loud, but it looks like this has a pretty hefty chunk of a linker in it. Have you experimented with ways to get the linker to do these relocations instead? Maybe something along the lines of having the toolchain emit a relocatable .o file as its "final" output using LLD `-r`, but with a linker script to assign the output sections, just not their addresses. Then analyzing that object to extract sizes, and running the linker again to relocate everything, including the ones dependent on the CD layout.

Not sure whether something like that would work, but if it could be made to, it should help make this easier to maintain; ideally something as small as a relatively thin wrapper around llvm command-line tools; it might not need a detailed understanding of ELF at all.

Reply:
I wouldn't say it's that hefty; compared to `elftocpm65`, the main difference is that it can also parse symbol names, and that I covered the sizes of all relocations "just in case".

I haven't experimented with getting the linker to do it much for the following reasons:

* I think `-r` would make it much more complicated, if not impossible, to assign things to absolute locations within a bank using a custom link script - binaries for the PC Engine CD are not unlike ROM images, just loaded at specific offsets in the 21-bit banked address space, and with the current approach, PCE CD link scripts are mostly similar in structure to existing PC Engine link scripts.
* How would the ability to link to another object's symbol be resolved?
  * Link them all together: this would cause annoying address space and symbol name overlaps. They are distinct programs, after all, and - more frustratingly - you may want to link to program B's `_start` address from program A (say, a game intro -> main game situation). An alternate solution of adding headers in front of every binary would open up the question of where to store them during loading - the CD has 2KB sectors, so a header would need to be padded to 2KB, while the unexpanded PCE CD has 72KB RAM total.
  * Provide a giant list of symbol definitions: you'd still need to generate addresses somehow before the final linking stage, in that case!
  
Do you see a way to accomplish this, while keeping a similar level of flexibility, which I don't?

Reply:
> Do you see a way to accomplish this, while keeping a similar level of flexibility, which I don't?

Nope, I just wanted to prod for alternatives. Having this depend on the full list of relocs means we'd need to keep this in sync with the linker and MC layer, which adds a bit of maintenance overhead to otherwise unrelated code. If there's any way around that that isn't too much worse, that would be desirable, but if there's not, there's not.

Elfcpm65 was a bit of a special case, in that it supports a set of relocs, and those are the relocs it supports. It's converting from one relocatable object file format to another, while this is acting a bit more.... linkey

Reply:
It doesn't have to depend on the full list of relocations; I just saw no reason not to add all of the ones supported so far (with the exception of special-case ones like `PCREL` or `ASCIIZ`), given the cost is one line per relocation.

Reply:
That's cool then, if the intent is to support a subset of available relocs for those outstanding symbols. I think the most likely future relocs would be related to PIC when and if iget around to it, so that all should be fairly separable

Reply:
I've added a comment, as to make it less confusing to future onlookers.